### PR TITLE
[W5.4][T14-A3]Lu Lechuan

### PR DESCRIPTION
**Pull request by LuLechuan**
Add a new Command - Undo: Undo either deletion or addition
Updated User Guide
Updated Test for undoing deletion

### CODE REVIEW

File: src/seedu/addressbook/commands/DeleteCommand.java [hunk 1]
```diff
@@ -29,6 +31,8 @@ public DeleteCommand(int targetVisibleIndex) {
     public CommandResult execute() {
         try {
             final ReadOnlyPerson target = getTargetPerson();
+            lastDeleted = (Person)target;
+            lastAdded = null;
```
Comment (majunting) [inline src/seedu/addressbook/commands/DeleteCommand.java]: maybe you can achieve both undo adding & undo deleting if you do not assign the lastAdded back to null here (and the lastDeleted in AddCommand class

**Comment by keloysiusmak** [comment]
Small point but maybe you want to have your pull request comments in present tense

File: src/seedu/addressbook/commands/UndoCommand.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        lastDeleted = _lastDeleted;
+    }
+
+    @Override
+    public CommandResult execute() {
+        if (lastAdded == null && lastDeleted == null){
+            return new CommandResult(MESSAGE_NOTHING_TO_UNDO);
+        } else if(lastAdded != null){
+            //System.out.println(addressBook.getAllPersons().immutableListView().size());
+            // DeleteCommand undoAdd = new DeleteCommand(addressBook.getAllPersons().immutableListView().size());
```
Comment (ang-zy) [inline src/seedu/addressbook/commands/UndoCommand.java]: Don't comment out code that you don't use. Delete it. If you really need it back at some point of time, you can always do a git reverse

File: test/expected.txt [hunk 3]
```diff
@@ -138,6 +140,23 @@
...
+|| ===================================================
+|| Enter command: || [Command entered:    list]
+|| 	1. Adam Brown Phone: 111111 Email: adam@gmail.com Address: 111, alpha street Tags: 
+|| 
+|| 1 persons listed!
+|| ===================================================
+|| Enter command: || [Command entered:    undo]
+|| Undo Addition successfully
+|| ===================================================
+|| Enter command: || [Command entered:    clear]
```
Comment (ang-zy) [inline test/expected.txt]: Why would you clear your addressbook after undoing a command? Does it help you test your undo command functionality?

File: test/expected.txt [hunk 4]
```diff
@@ -175,28 +194,28 @@
 || The person index provided is invalid
 || ===================================================
 || Enter command: || [Command entered:  view 1]
-|| Viewing person: Adam Brown Phone: 111111 Email: adam@gmail.com Address: 111, alpha street Tags: 
+|| Person could not be found in address book
```
Comment (ang-zy) [inline test/expected.txt]: What are the tests here supposed to check for? Have they achieved their purpose?

File: test/input.txt [hunk 5]
```diff
@@ -55,6 +55,17 @@
   add Esther Potato p/555555 e/esther@not.a.real.potato pa/555, epsilon street t/tubers t/starchy
 
 ##########################################################
+# test undo command, setup state for futuer tests
+##########################################################
+
+  # should add correctly and list non private information
+    clear
```
Comment (ang-zy) [inline test/input.txt]: Does this chunk of tests affect the rest of your tests?

File: src/seedu/addressbook/commands/UndoCommand.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    public void setLastDeleted(Person _lastDeleted){
+        lastDeleted = _lastDeleted;
+    }
+
+    @Override
+    public CommandResult execute() {
+        if (lastAdded == null && lastDeleted == null){
+            return new CommandResult(MESSAGE_NOTHING_TO_UNDO);
+        } else if(lastAdded != null){
```
Comment (ang-zy) [inline src/seedu/addressbook/commands/UndoCommand.java]: Your undo logic might be flawed. What happens when you add -> delete -> undo?

**Pull request closed without merging**